ffmpeg/lpms_ffmpeg.c

```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
   if (md) av_dict_free(&md);
   return ret == AVERROR_EOF ? 0 : ret;
 }
+
+//
+// Transcoder
+//
+
+static void free_filter(struct filter_ctx *filter)
+{
+  if (filter->frame) av_frame_free(&filter->frame);
+  if (filter->graph) avfilter_graph_free(&filter->graph);
+}
+
+static void free_output(struct output_ctx *octx)
+{
+  if (octx->oc) {
+    if (!(octx->oc->oformat->flags & AVFMT_NOFILE) && octx->oc->pb) {
+      avio_closep(&octx->oc->pb);
+    }
+    avformat_free_context(octx->oc);
+  }
+  if (octx->vc) avcodec_free_context(&octx->vc);
+  if (octx->ac) avcodec_free_context(&octx->ac);
+  free_filter(&octx->vf);
+  free_filter(&octx->af);
+}
+
+
+static int open_output(struct output_ctx *octx, struct input_ctx *ictx)
+{
+#define em_err(msg) { \
+  if (!ret) ret = -1; \
+  fprintf(stderr, msg); \
+  goto open_output_err; \
+}
+  int ret = 0;
+  AVOutputFormat *fmt = NULL;
+  AVFormatContext *oc = NULL;
+  AVCodecContext *vc  = NULL;
+  AVCodecContext *ac  = NULL;
+  AVCodec *codec      = NULL;
+  AVStream *st        = NULL;
+
+  // open muxer
+  fmt = av_guess_format(NULL, octx->fname, NULL);
+  if (!fmt) em_err("Unable to guess output format\n");
+  ret = avformat_alloc_output_context2(&oc, fmt, NULL, octx->fname);
+  if (ret < 0) em_err("Unable to alloc output context\n");
+  octx->oc = oc;
+
+  if (ictx->vc) {
+    codec = avcodec_find_encoder_by_name("libx264"); // XXX make more flexible?
+    if (!codec) em_err("Unable to find libx264");
+
+    // open video encoder
+    // XXX use avoptions rather than manual enumeration
+    vc = avcodec_alloc_context3(codec);
+    if (!vc) em_err("Unable to alloc video encoder\n"); // XXX shld be optional
+    octx->vc = vc;
+    vc->width = av_buffersink_get_w(octx->vf.sink_ctx);
```

Comment:
Anywhere I can learn about the buffer source / buffer sink API?

Reply:
Probably the canonical source for those specific functions is the header documentation, the comments are fairly thorough: https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/buffersink.h